### ISSUE TITLE
Expand modulemap to cover all standalone headers.

### ISSUE THREAD

**Issue by Teemperor** [Teemperor/boost-compile#1]
So far I wrote a simple modulemap that I placed in the boost directory: https://github.com/Teemperor/boost-compile/blob/master/inc/boost/module.modulemap

This simplicity brings some problems with it. For example all the headers in the subdirectories are currently included in all the modules that use them. E.g. the `array.hpp` and the `circular_buffer.hpp` both include `boost/detail/iterator.hpp`. This header is now duplicated in both modules which makes things complicated and less effective. So this header needs to be added to the modulemap.

However, we can't just add all headers to the modulemap, but only those headers which are standalone and can be parse on their own. So for example all `*.ipp` files are supposed to be included textually, not as a module.

A simple, but probably not perfect way to check is to take each header and try to parse it's on its own. If clang can parse it, it's intended to be a hopefully a standalone header. Same goes if it doesn't have header guards.

**Comment by Teemperor** [comment]
On another note: Module files should never reach 512 MB, otherwise clang's offsets in the file overflow and things will behave strangely. We should also modify the modulemap to make sure this never happens. I'm not sure if hit happens with the current modulemap, so it would be good to check this.

**Comment by Teemperor** [comment]
Also, we probably want to test this with a modulemap for STL. Easiest way to do this is drop this into your STL include folder: https://github.com/Teemperor/ClangModulesCMake/blob/master/files/stl14.modulemap (Use -Rmodule-build to test this)

Boost uses STL, so with a STL modulemap we should avoid some decl merging issues in there.

**Comment by yamaguchi1024** [comment]
https://gist.github.com/de652697cb57d83d01109d8145f3c8d4
I just generated this from shellscript, but is this in correct format? I was not sure what you were talking about "standalone headers".

**Comment by Teemperor** [comment]
Can you push the shell script? Feel free to dump these things into the repo.

So standalone headers are headers that can be parsed on their own. For example this is not a standalone header because if you try to parse on its own, clang will say that it doesn't know what std::vector is:

```C++
std::vector<int> listOfPrimeNumbers();
```

You can only use it by using this header like that:
```C++
#include <vector>
#include "ListOfPrime.h" // this is he header above
```

This would be the standalone version of the header:

```C++
#include <vector>
std::vector<int> listOfPrimeNumbers();
```

So standalone only means that it supposed to be parseable on its own. If we can't parse it own its own, than we can't create a module for this header (because to create a module, we need to parse the header into an AST, and then serialize this AST, and getting an AST requires that we can parse it).

**Comment by Teemperor** [comment]
Things are obviously not always that easy. So for example this header parses on its own, but it's not supposed to parse on its own:
```C++
// Define TYPE macro before including this header if you need something else than int.
#ifndef TYPE
#define TYPE int
#endif
struct Container {
   TYPE data[10];
};
```

If we turn this into a module, then things might break in very cryptic ways as we now always have type int in here.

**Comment by Teemperor** [comment]
And yes, the format looks correct. However, you could merge some modules like this:
```C++
module TopModule {
  module SubModule1 { header "m1.h" export * }
  module SubModule2 { header "m2.h" export * }
}
```
This will cause clang to compile both modules together into a single `.pcm` file. There are some advantages in this, mostly that if m1.h and m2.h include each other in a cycle, then clang can handle this nicely (as long as it would also compile fine without modules if you include both m1.h and m2.h).

However, if you merge too many modules into one, then the PCMs get too big and hit the 512 MB limit. Also, if you make too big modules then it takes clang a long time to create a PCM and maybe the user just wants to use a single header from boost and not all of it.

I think a good split up could be that each boost subdirectory maybe is one module?

**Comment by Teemperor** [comment]
I just noticed that we probably shouldn't have "/" in module names. This actually causes clang to create subdirectories in the PCM directory :/

**Comment by yamaguchi1024** [comment]
I used this shell script(zsh):
```
find . -name "*.hpp" | sed "s/\.\///g" | while read line
do
clang++ -fsyntax-only $line
if [ $? -ne 0 ]; then
  continue
fi
echo "module \"$line\" { header \"$line\" export * }"
done > ~/dev/module.modulemap
```
and got 10000+ lines of modulemap. However when I look at clang and llvm's modulemap, it's like 300 lines maybe because they're using umbrella.
200KB+ modulemap doesn't look like very smart, I wonder if there's a way to reduce this size without checking each file manually.

**Comment by Teemperor** [comment]
Yeah, but boost is also a thousand times bigger than the LLVM headers. if the modulemaps works, can you just make a PR? We can reduce it later if it works.

**Comment by yamaguchi1024** [comment]
How can I check if it works or not? We have thousands of errors anyway, so I'm not sure what's the correct behavior.

**Comment by Teemperor** [comment]
Well, the script needs two fixes: First you can't just use clang on it, because you want to filter out all headers that have things like main and don't have header guards. The best test I can think of is

```C++
#include <HEADER>
#include <HEADER> // Include twice, if HEADER has no header guards it shouldn't be in the modulemap and this will error in this case to filter it out.

int main() {} // if HEADER or a included header from HEADER defines main, this creates an error as we dont wan't those header in our modulemap.
```

and then replace HEADER with the path and then try to compile with clang.

And then we get a modulemap that should hopefully bring us closer to getting boost to compile. If we get compilation errors like cyclic dependency errors from clang then we're on the right track. But it's otherwise hard to say if the modulemap is correct or not at this point :)

**Comment by yamaguchi1024** [comment]
I saved your code as main.cpp, and changed script as below:
```
find . -name "*.hpp" | sed "s/\.\///g" | while read line
do
sed "s#HEADER#$line#g" ~/dev/main.cpp > /tmp/tmp.cpp
clang++ -fsyntax-only -I ~/boost-compile/inc/boost /tmp/tmp.cpp
if [ $? -ne 0 ]; then
  continue
fi
echo "module \"$line\" { header \"$line\" export * }"
done > ~/dev/modulemap
```
Generated modulemap size is far smaller and looks less ugly. Thanks!

**Comment by Teemperor** [comment]
Two points: Clang is by default in c++03, so all the things in boost that have c++11 and higher requirements won't parse with the script. Also are the clang include paths correct? Because I think you need to have `-I/boost-compile/inc` as boost includes everything with `#include "boost/header.hpp"` and not just `#include "header.hpp"`

**Comment by Teemperor** [comment]
I think it makes sense that we use c++14 for compilation testing. C++17 is maybe a bit too fresh out of the box.

**Comment by Teemperor** [comment]
I think it would make sense if the script makes records of the stderr output of clang. We probably have to recheck those errors in the future when we have to see why a certain header is not in the modulemap.

I also thought about having `-Werror` in the clang invocation, because if a header emits warnings it could be a sign that we shouldn't include it directly. We probably have to check how many headers actually emit warnings, so maybe we just keep this in mind for the future. Let's first get a somewhat logical modulemap :)

**Comment by yamaguchi1024** [comment]
Ah, I think clang was reading from my system boost, not from this repository.
```
find . -name "*.hpp" | sed "s/\.\///g" | while read line
do
sed "s#HEADER#$line#g" ~/dev/main.cpp > /tmp/tmp.cpp
clang++ -fsyntax-only -I ~/boost-compile/inc -std=c++14 /tmp/tmp.cpp 2>> ~/dev/module_error
if [ $? -ne 0 ]; then
  continue
fi
echo "$line"
done > ~/dev/modulemap
```

**Comment by Teemperor** [comment]
Ok, running the generator right now with this modified script:
```bash
# Clear output file
echo "" > ~/boost-compile/module_error

# Find all headers and remove the /inc part
find . -name "*.hpp" | sed "s/\.\///g" | cut -c 5- | while read line
do

sed "s#HEADER#$line#g" ~/boost-compile/main.cpp > /tmp/tmp.cpp

echo "#####################################################" >> ~/boost-compile/module_error
echo "HEADER: $line" >> ~/boost-compile/module_error
echo "#####################################################" >> ~/boost-compile/module_error

clang++ -Werror -I ~/boost-compile/inc -std=c++14 /tmp/tmp.cpp 2>> ~/boost-compile/module_error
if [ $? -ne 0 ]; then
  (>&2 echo "FAIL: $line")
  continue
fi
(>&2 echo "GOOD: $line")
echo "$line"
done > ~/boost-compile/modulemap
```

**Comment by Teemperor** [comment]
This should now produce a list of headers that compile at the moment. We still need to merge them based on subdirectory or something like that.

**Comment by Teemperor** [comment]
I uploaded the list of headers from your script here: https://teemperor.de/pub/module_headers

**Comment by yamaguchi1024** [comment]
http://yamaguchi1024.dip.jp/pub/module.modulemap
How about this?

**Comment by Teemperor** [comment]
Looks better, can you implement this as part of the script in #4 so that we can automatically generate this? (Because that way we don't have to go through this same hassle with other boost versions in the future). Feel free to use python or so if that makes the script simpler.

By the way, the paths to the headers are in your file just filenames, not the relative paths to the modulemap file.

**Comment by yamaguchi1024** [comment]
I used ruby to generate that modulemap, but maybe I should change it to shellscript.

What do you mean "the paths to the headers are in your file just filenames"? I think I don't have any paths in that modulemap

**Comment by Teemperor** [comment]
You can also just add the ruby script. It should just be readable and have no hardcoded paths :)

I mean that `module "list_inserter.hpp" { header "list_inserter.hpp" export * }` should be `module "list_inserter.hpp" { header "assign/list_inserter.hpp" export * }`. The `header` field is the relative path to the header from the directory in which the modulemap is.

**Comment by yamaguchi1024** [comment]
Ah, that's true.. Thanks

**Comment by yamaguchi1024** [comment]
Updated the link, and this is the ruby script I was using.
```
ptop = ""
File.open("mod","r").read.each_line do |line|
  sline = line.split(File::SEPARATOR)
  top = sline[0]
  if sline.count == 1 then
    if ptop != "" then
      puts "}"
    end
    print "module \"" + top.strip + "\" { header \"" + top.strip + "\" export * }\n"
    ptop = ""
    next
  end
  if ptop != top then
    if ptop != "" then
      puts "}"
    end
    puts "module " + top + " {"
  end
  ptop = top
  print "  module \"" + sline.last.strip + "\" { header \"" + line.strip + "\" export * }\n"
end
puts "}"
```

**Comment by Teemperor** [comment]
Cool, we might need to fix the module names though. For example `named_condition_any.hpp` is twice in `interprocess` module. The best way is to take the realtive path, replace `/` with `_` and remove the `.hpp` suffix. So instead of `  module "named_condition_any.hpp" { header "interprocess/sync/named_condition_any.hpp" export * }` we would have `  module "interprocess_sync_named_condition_any." { header "interprocess/sync/named_condition_any.hpp" export * }`

**Comment by yamaguchi1024** [comment]
https://gist.github.com/yamaguchi1024/c4dd7c0c8856639249d9cbc5cd1b33bd
```ruby
ptop = ""
File.open("module_header","r").read.each_line do |line|
  # delete "boost/ and parse each line separated by "/""
  sline = line.split(File::SEPARATOR)[1..-1]
  top = sline[0]
  # if the hpp file was in toplevel directory
  if sline.count == 1 then
    if ptop != "" then
      puts "}"
    end
    print "module \"" + top.strip[0..-5] + "\" { header \"" + top.strip + "\" export * }\n"
    ptop = ""
    next
  end
  if ptop != top then
    if ptop != "" then
      puts "}"
    end
    puts "module " + top + " {"
  end
  ptop = top
  print "  module \"" + sline.join("_").strip[0..-5] + "\" { header \"" + sline.join("/").strip + "\" export * }\n"
end
puts "}"
```

**Comment by yamaguchi1024** [comment]
Is this ok?
```
module "align" { header "align.hpp" export * }
module align {
  module "align_is_aligned" { header "align/is_aligned.hpp" export * }
}
```

**Comment by Teemperor** [comment]
Hmm, only problem I see is that we now have two `align` modules which probably breaks clang. Can we move the `align.hpp` into the `align` module? I don't think all top level headers have a corresponding directory, but many have.

**Comment by yamaguchi1024** [comment]
Is it not allowed to write modules separately? For example:
```
module a {
  module something...
}
...
module a {
  module something..
}
```

**Comment by Teemperor** [comment]
I don't think so, it's probably hard to know if this happened on purpose or not.

**Comment by yamaguchi1024** [comment]
I think there's no error for modulemap itself now.
https://gist.github.com/yamaguchi1024/0247e82bd54ad05cca01a32502ce46a5
```ruby
#!/usr/bin/ruby

modules = Hash.new
File.open("module_header","r").read.each_line do |line|
  # delete "boost/ and parse each line separated by "/""
  sline = line.split(File::SEPARATOR)[1..-1]
  top = sline[0]
  if sline.count == 1 then
    top = top.strip[0..-5]
  end
  # if the hpp file was in toplevel directory
  if !modules[top] then
    modules[top] = Array.new
  end
  modules[top].push(sline)
end

modules.each{|key, array|
  print "module " + key + " {\n"
  array.each do |line|
    print "  module \"" + line.join("__").strip[0..-5] + "\" { header \"" + line.join("/").strip + "\" export * }\n"
  end
  print "}\n"
}
```

**Comment by yamaguchi1024** [comment]
Now I have a lot of `declaration of '<something>' must be imported from module '<something>' before it is required` error

**Comment by Teemperor** [comment]
Can you make a PR of the modulemap you have that I can reproduce them? Just replace the old modulemap with your new one.

**Comment by vgvassilev** [comment]
@yamaguchi1024 that in many cases means that the header is not standalone. Would adding the relevant `#include` to the header file solve the issue?

**Comment by Teemperor** [comment]
Well, our test case for all headers ensures that they are standalone. So I fear those are visibility bugs. Are you using trunk clang/5.0 or still 4.0?

**Comment by yamaguchi1024** [comment]
I made a PR, and I'm using clang/5.0

**Comment by Teemperor** [comment]
can you double check it's the latest clang? For me I have cyclic dependency errors with the modulemap (which we can solve). Missing imports don't occur for me.

**Comment by Teemperor** [comment]
Also, post the errors from clang so that we can check if it's maybe just a missing modulemap in an STL folder or so

**Comment by yamaguchi1024** [comment]
https://gist.github.com/yamaguchi1024/5fc2d52a36e6eb6e4a30b5f4bcdd6920

**Comment by vgvassilev** [comment]
Yeah, that's what I suspected. It's libc++ and probably the fact it doesn't support -fmodules-local-submodule-visibility...

**Comment by yamaguchi1024** [comment]
I'll try on different OS, thanks

**Comment by yamaguchi1024** [comment]
Now I have "cyclic dependency" errors. Is that what you are talking about yesterday?

**Comment by Teemperor** [comment]
Yes! This means the module itself are doing fine, but the modulemap needs tuning. The problem is that if you have two problems that include each other then clang can't build this. The way to solve this is by splitting out the modules to resolve the cycle. E.g.
``` C++
module system {
   module console { header "console.h" export *} // uses logger.h from [logging]
   module file { header "file.h" export *}
}

module logging {
  module logger { header "logger.h" export * }
  module colorizer { header "colorizer.h" export * } // uses file.h from [system]
}
```
 Here system and logging have a cyclic dependency between them. You can resolve this by doing:

``` C++
module system {
   module console { header "console.h" export *} // uses logger.h from [logging]
   module file { header "file.h" export *}
}

module logger { header "logger.h" export * }
module colorizer { header "colorizer.h" export * } // uses file.h from [system]
```

**Comment by Teemperor** [comment]
I have some tool to solve this, it should also work for boost. Give me a bit to see if it works as is for boost.

**Comment by Teemperor** [comment]
The tool is here: https://github.com/Teemperor/circle-break

**Comment by yamaguchi1024** [comment]
Fixed some cycle and send PR, but now I get "could not build" errors and then segfaults instead

fatal error: could not build module 'type_traits'

**Comment by yamaguchi1024** [comment]
#11

**Comment by yamaguchi1024** [comment]
^ mistyped, sorry!
